### ISSUE TITLE
run error: emqx_vm:get_memory().

### ISSUE THREAD

**Issue by noerl** [emqx/emqx#3767]
run error:
emqx_vm:get_memory().  

**Environment**:  

- EMQ X version (e.g. `emqx_ctl status`):  4.2.1
- Hardware configuration (e.g. `lscpu`):
- OS (e.g. `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Erlang/OTP version (in case you build emqx from source code):  OTP-23
- Others: 

**What happened and what you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

(emqx@127.0.0.1)1> emqx_vm:get_memory().
** exception error: bad argument
     in function  element/2
        called as element(2,false)

emqx/src/emqx_vm.erl, line 313
emqx/src/emqx_vm.erl, line 295
emqx/src/emqx_vm.erl, line 272
    

**Anything else we need to know?**:

    %% OTP-22 introduces carrier migrations across types, and OTP-23 changes the
    %% format of data reported to be a bit richer; however it's not compatible
    %% with most calculations made for this library.
    %% So what we do here for `blocks' is merge all the info into the one the
    %% library expects (`blocks' and `blocks_size'), then keep the original
    %% one in case it is further needed.
    %% There were further changes to `mbcs_pool' changing `foreign_blocks',
    %% `blocks' and `blocks_size' into just `blocks' with a proplist, so we're breaking
    %% up to use that one too.
    %% In the end we go from `{blocks, [{Alloc, [...]}]}' to:
    %%  - `{blocks, ...}' (4-tuple in mbcs and sbcs, 2-tuple in mbcs_pool)
    %%  - `{blocks_size, ...}' (4-tuple in mbcs and sbcs, 2-tuple in mbcs_pool)
    %%  - `{foreign_blocks, [...]}' (just append lists =/= `Alloc')
    %%  - `{raw_blocks, [...]}' (original value)

**Comment by noerl** [comment]
reference: https://github.com/ferd/recon/blob/master/src/recon_alloc.erl
fix:  
emqx_vm.erl

`allocators() ->
    UtilAllocators = erlang:system_info(alloc_util_allocators),
    Allocators = [sys_alloc, mseg_alloc|UtilAllocators],
    [{{A, N},lists:sort(format_alloc(A, proplists:delete(versions, Props)))} ||
        A <- Allocators, Allocs <- [erlang:system_info({allocator, A})],
            Allocs =/= false, {_, N, Props} <- Allocs].

format_alloc(Alloc, Props) ->
    %% {versions,_,_} is implicitly deleted in order to allow the use of the
    %% orddict api, and never really having come across a case where it was
    %% useful to know.
    [{K, format_blocks(Alloc, K, V)} || {K, V} <- lists:sort(Props)].

format_blocks(_, _, []) ->
    [];
format_blocks(Alloc, Key, [{blocks, L} | List]) when is_list(L) ->
    %% OTP-22 introduces carrier migrations across types, and OTP-23 changes the
    %% format of data reported to be a bit richer; however it's not compatible
    %% with most calculations made for this library.
    %% So what we do here for `blocks' is merge all the info into the one the
    %% library expects (`blocks' and `blocks_size'), then keep the original
    %% one in case it is further needed.
    %% There were further changes to `mbcs_pool' changing `foreign_blocks',
    %% `blocks' and `blocks_size' into just `blocks' with a proplist, so we're breaking
    %% up to use that one too.
    %% In the end we go from `{blocks, [{Alloc, [...]}]}' to:
    %%  - `{blocks, ...}' (4-tuple in mbcs and sbcs, 2-tuple in mbcs_pool)
    %%  - `{blocks_size, ...}' (4-tuple in mbcs and sbcs, 2-tuple in mbcs_pool)
    %%  - `{foreign_blocks, [...]}' (just append lists =/= `Alloc')
    %%  - `{raw_blocks, [...]}' (original value)
    Foreign = lists:filter(fun({A, _Props}) -> A =/= Alloc end, L),
    Type = case Key of
        mbcs_pool -> int;
        _ -> quadruple
    end,
    MergeF = fun(K) ->
        fun({_A, Props}, Acc) ->
            case lists:keyfind(K, 1, Props) of
                {K,Cur,Last,Max} -> {Cur, Last, Max};
                {K,V} -> Acc+V
            end
        end
    end,
    %% Since tuple sizes change, hack around it using tuple_to_list conversion
    %% and set the accumulator to a list so it defaults to not putting anything
    {Blocks, BlocksSize} = case Type of
        int ->
            {{blocks, lists:foldl(MergeF(count), 0, L)},
             {blocks_size, lists:foldl(MergeF(size), 0, L)}};
        quadruple ->
            {list_to_tuple([blocks | tuple_to_list(lists:foldl(MergeF(count), {0,0,0}, L))]),
             list_to_tuple([blocks_size | tuple_to_list(lists:foldl(MergeF(size), {0,0,0}, L))])}
    end,
    [Blocks, BlocksSize, {foreign_blocks, Foreign}, {raw_blocks, L}
     | format_blocks(Alloc, Key, List)];
format_blocks(Alloc, Key, [H | T]) ->
    [H | format_blocks(Alloc, Key, T)].
`

**Comment by HJianBo** [comment]
Hi, @noerl Thanks for your hint.

Full OTP-23 support will be implemented in emqx5.0, until then please compile emqx using OTP-22 :)

**Comment by k32** [comment]
This error has been fixed in the current version (testing on 4.3).

```
Erlang/OTP 23 [erts-11.1.7] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [hipe]

Eshell V11.1.7  (abort with ^G)
(emqx@10.0.0.5)1> emqx_vm:get_memory().
[{used,107323736},
 {allocated,146640896},
 {unused,39317160},
 {usage,0.7318813436600933},
 {total,118702352},
 {processes,45693144},
 {processes_used,45692080},
 {system,73009208},
 {atom,1499481},
 {atom_used,1468060}, 
 {binary,1167336},
 {code,34349708},
 {ets,7350288}]
```
